src/main/java/htsjdk/samtools/read/Read.java

```diff
@@ -0,0 +1,1438 @@
+     *
+     * <p>Default implementation returns {@code getBases()[i]}.
+     * Subclasses may override to provide a more efficient implementation.
+     *
+     * @return base at index i.
+     *
+     * @throws IndexOutOfBoundsException if i is negative or of i is not smaller than the number
+     * of bases (as reported by {@link #getLength()}. In particular, if no sequence is present.
+     */
+    default byte getBase(final int i){
+        return getBases()[i];
+    }
+
+    /**
+     * Gets all the bases in the read as a String of ACGTN=
+     *
+     * <p>Equivalent to <b>SEQ</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Default implementation uses {@code getBases()}.
+     * Subclasses may override to provide a more efficient implementation.
+     *
+     * @return read sequence as a string ; {@link ReadConstants#NULL_SEQUENCE_STRING} if the read is empty.
+     */
+    default String getBasesString() {
+        return isEmpty() ? ReadConstants.NULL_SEQUENCE_STRING : StringUtil.bytesToString(getBases());
+    }
+
+    /**
+     * Sets the sequence bases.
+     *
+     * <p>Equivalent to <b>SEQ</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * @param bases read sequence as ASCII bytes ACGTN=; {@link ReadConstants#NULL_SEQUENCE} if no sequence is present.
+     *
+     * @throws IllegalArgumentException if the bases are null.
+     */
+    void setBases(final byte[] bases);
+
+    /**
+     * Gets the base qualities, as binary PHRED scores (not ASCII).
+     *
+     * <p>Equivalent to <b>QUAL</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>This method should make a defensive copy of the bases array before returning it to allow
+     * modification of the returned array without effects on the {@link Read}.
+     *
+     * @return base qualities; {@link ReadConstants#NULL_QUALS} if base qualities are not present.
+     */
+    byte[] getBaseQualities();
+
+    /**
+     * Gets the number of base qualities in the read sequence.
+     *
+     * <p>Equivalent to <b>QUAL</b> in the <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM specifications</a>.
+     *
+     * <p>Default implementation calls {@code getBaseQualities().length}.
+     * Subclasses may override to provide a more efficient implementation.
+     *
+     * @return number of base qualities in the read sequence.
+     */
+    default int getBaseQualityLength(){
+        return getBaseQualities().length;
+    }
+
+    /**
+     * Checks if the read has qualities.
+     *
+     * <p>Default implementation returns {@code getBaseQualityLength() != 0}.
+     *
+     * @return {@code true} if the read has qualities; {@code false} otherwise.
+     */
+    default boolean hasQualities() {
+        return getBaseQualityLength() != 0;
```

Comment:
If implementation details are maintained, this one will be more efficient if the sub-class access directly the length; otherwise, this will do a defensive copy...